Fix v7.7.x Ritual Tracking Post v7.6.0 Rebase

derekpierre opened the pull request:
**Type of PR:**
- [ ] Bugfix
- [ ] Feature
- [ ] Documentation
- [x] Other

**Required reviews:** 
- [ ] 1
- [X] 2
- [ ] 3

**What this does:**
Updates needed post rebase of v7.7.x over `main` (just released v7.6.0) that included handover changes. The ritual tracker was significantly refactored in v7.7.x before the rebase as part of #3599 , and so there were some conflicts/issues with rebasing over v7.6.0. 

This PR aims to solve that problem after a basic rebase that caused ritual tracking to no longer work properly.

**Issues fixed/closed:**
> - Fixes #...

**Why it's needed:**
> Explain how this PR fits in the greater context of the NuCypher Network.
> E.g., if this PR address a `nucypher/productdev` issue, let reviewers know!

**Notes for reviewers:**

Key changes include:
- Refactored participation state management to handle both DKG and handover events separately
- Fixed participation state storage to use string keys consistently
- Updated test methods to use new participation state value retrieval methods

derekpierre committed:
Modify how state is managed for the ritual trackers. Handover can change participation so we need to update our cached previous participation state before using it.

derekpierre committed:
Modify how state is managed for the dkg tracker:
1) Handover and dkg participation states are tracked separately via using separate identifiers
2) Since handover still requires other nodes not in the handover process to act, participating involves both involved in handover, or a current participant in ritual
3) Handover can change participation so we need to update our cached participation state before using it.
4) Track HandoverCanceled as well so that participation can be updated accordingly.
5) Ritual id identifier for dkg states should be a string not a number.

derekpierre committed:
Adjust signing ritual tracker to accomodate new paradigm where participation can change.
Cohort id identifier used for participation states should be a string not an int.

derekpierre committed:
Update tests to use updated function and to account for Handover participation state being separate from dkg participation state.

derekpierre committed:
Account for extra fields in other Handover events.

derekpierre committed:
Use string version of ritual id when obtaining participation states straight from the TTL cache used by the RitualTracker base class.

derekpierre committed:
Add dev newsfragment for #3638.

derekpierre committed:
Comment cleanup for participation state classes.

KPrasch commented on nucypher/blockchain/eth/trackers/dkg.py:
```diff
@@ -25,10 +28,15 @@ def __init__(
+        Participation state for Handover rituals.
+        """
+
+        PREFIX = "handover-"
```
so any event with this prefix will be processed?

derekpierre replied:
It's part of the key for state tracking.

We use the same tracker for tracking DKG events and Handover events. If we tried to track state using the same state object, it just complicates the logic. So instead we track state for those sets pf events separately. See _particiaptionState (TTL cache).

For state checking regarding DKG events (StartRitual, StartAggregationRound, EndRitual) just use ritualID as the key to store/get the state eg. "12", but for handover state checking regarding Handover events (HandoverRequested, HandoverTranscriptPosted, HandoverFinalized, HandoverCanceled) the key is `handover-rituaID` eg. "handover-12".

It's basically a way of making sure that DKG state is always separate from Handover state even for the same ritual.

The states are needed to prevent us from making a bunch of RPC calls during DKG and Handover Rituals.

derekpierre committed:
Use _get_identifier function for determining key in participation states dictionary for DkgRitualTracker instead of assuming it will always be the string version of ritual id (which it is, but if we ever change that to use a prefix or some other format like handover did then the tests would fail and need to be updated).

derekpierre committed:
Add tests for SigningRitualTracker to improve code coverage in this area. It mimics tests from the DkgRitualTracker but specific to signing rituals.

derekpierre commented on tests/acceptance/actors/test_signing_ritual_tracker.py:
```diff
@@ -0,0 +1,808 @@
+            "args": AttributeDict(
+                {
+                    "cohortId": cohort_id_3,
+                    "chainIid": chain_id,
```
✔️

derekpierre commented on tests/acceptance/actors/test_signing_ritual_tracker.py:
```diff
@@ -0,0 +1,808 @@
+    initiate_signing_cohort = agent.contract.events.InitiateSigningCohort()
+    args_dict = {
+        "cohortId": cohort_id_1,
+        "chain": chain_id,
```
✔️

manumonti commented on nucypher/blockchain/eth/trackers/dkg.py:
```diff
@@ -62,47 +72,64 @@ def __init__(
 
-            if event_type == self.contract.events.HandoverFinalized:
-                # pruning metadata is sufficient when Handover is finalized;
+            if event_type in [
+                self.contract.events.HandoverFinalized,
+                self.contract.events.HandoverCanceled,
+            ]:
+                if (
+                    event_type == self.contract.events.HandoverCanceled
+                    and event.args.incomingParticipant == self.operator.checksum_address
+                ) or (
+                    event_type == self.contract.events.HandoverFinalized
+                    and event.args.departingParticipant
+                    == self.operator.checksum_address
+                ):
+                    # if handover is canceled/finalized we need to reset the participation state
+                    participation_state.participating = False
+
+                # pruning metadata/cleanup is sufficient when Handover is finalized or canceled;
                 # no further action required
                 return False
```
Pretty sure I'm misunderstanding the handover logic here, but is it actually possible to have a HandoverCanceled or HandoverFinalized but with the node still participating of it?

In other words: is this the same that just


```suggestion
            if event_type in [
                self.contract.events.HandoverFinalized,
                self.contract.events.HandoverCanceled,
            ]:
                # pruning metadata/cleanup is sufficient when Handover is finalized or canceled;
                # no further action required
                return False
```

derekpierre replied:
> is it actually possible to have a HandoverCanceled or HandoverFinalized but with the node still participating of it?

Yes. This is the case for nodes that are in the cohort and not part of handover (neither departing or incoming participant). Participation pertains to the ~~signing~~ cohort not the handover.

> In other words: is this the same that just

Not quite.

The underlying logic for participation after the handover is finished is
1. If handover was canceled then:
a) if the node was the incoming participant, then it is not participating in the cohort
b) if the node was the departing participant, then it is still participating in the cohort (stays the same since set before)
c) if the node was neither then its participation remains the same (no change)
2. If handover was finalized then:
a) if the node was the departing participant, then it is no longer participating in that cohort
b) if the node was the incoming participant, then it is now participating in that cohort (stays the same since set before)
c) if the node was neither then its participation remains the same (no change)

You are right for cases b) and c) in both 1) and 2), but it is the a) case for 1) and 2) what this extra check in the code is performing.

manumonti replied:
oooh, got it. Yes, as you are pointing, I thought participant meant that you are a participant of the handover instead of the ritual. Now it's clear, thanks.

derekpierre replied:
I'll update some of the class documentation to help clarify that. Based on the naming I can see where the confusion came from.

derekpierre replied:
@manumonti, added https://github.com/nucypher/nucypher/pull/3638/commits/550dcda3909e2576379a6c8c51e2b62d1c5f3919.

manumonti reviewed (approved):
LGTM! although I have a suggestion

derekpierre committed:
Updated documentation and rename participation state object used for handovers to clarify that it pertains to cohort participation and not participation in the handover process itself.

Co-authored-by: Manuel Montenegro <manuel@nucypher.com>

cygnusv commented on nucypher/blockchain/eth/trackers/dkg.py:
```diff
@@ -62,47 +72,66 @@ def __init__(
+        if event_type in self.handover_event_types:
+            return f"{self.CohortParticipationStateDuringHandover.PREFIX}{event.args.ritualId}"
+        else:
+            return str(event.args.ritualId)
```
Shouldn't DKG participation state have a prefix going forward too (e.g, "DKG")? For consistency purposes, mainly. Not a hard comment, though.

derekpierre replied:
It can if we want. I just didn't since the dkg state was the first and default from before.

✔️ good to remain consistent, so added commit - https://github.com/nucypher/nucypher/pull/3638/commits/3587494e33d0975cffd84d3a2bf181951f8ecde2.

cygnusv commented on nucypher/blockchain/eth/trackers/dkg.py:
```diff
@@ -62,47 +72,66 @@ def __init__(
+                    event_type == self.contract.events.HandoverFinalized
+                    and event.args.departingParticipant
+                    == self.operator.checksum_address
+                ):
```
For increased readability, you could use `is_incoming_participant_in_handover` and `is_departing_participant_in_handover` from below here.

derekpierre replied:
✔️  - great catch

derekpierre committed:
Add participation state prefixes for Dkg and Signing states.
Add associated tests.
Update lingering test that didn't use proper state identifier.

derekpierre committed:
Improve readability of logic for handling HandoverFinalized/Canceled events.
